### PR TITLE
Instance: Allow storage volume config to be migrated

### LINKED ISSUE #10102: Instance volume migration does not transfer volume DB config
When doing a cross-pool copy of an instance, the storage volume DB record config is not transferred (unlike when doing a custom storage volume cross-pool copy).

E.g.

```
lxc storage create lvm1 lvm
lxc storage set lvm1 volume.block.filesystem=xfs
lxc storage create lvm2 lvm
lxc init images:ubuntu/focal c1 -s lvm1

lxc storage volume show lvm1 container/c1
config:
  block.filesystem: xfs
  block.mount_options: discard
description: ""
name: c1
type: container
used_by:
- /1.0/instances/c1
location: none
content_type: filesystem

lxc copy c1 c2 -s lvm2
lxc storage volume show lvm2 container/c2
config:
  block.filesystem: ext4
  block.mount_options: discard
description: ""
name: c2
type: container
used_by:
- /1.0/instances/c2
location: none
content_type: filesystem
```

The `block.filesystem` setting has not been preserved. 

Also snapshot storage records are not transferred for either instance or custom volume types.

- [x] Fix instance storage volume (and snapshots) config copying. (https://github.com/lxc/lxd/pull/10526).
- [x] Fix custom storage volume (and snapshots) config copying (https://github.com/lxc/lxd/pull/10423).

### PR DESCRIPTION
**Pull request by tomponline**
- Introduces index migration header support for instance migration.
- This allows the storage volume and storage volume snapshot config to be propagated to the target (Fixes #10102).
- Includes error message and error handling improvements in migration logic.
- Improves validation of source and target state before starting migration to ensure they are in consistent state.

Includes https://github.com/lxc/lxd/pull/10510

### CODE REVIEW

File: lxd/migrate_instance.go [hunk 1]
```diff
@@ -446,26 +463,24 @@ func (s *migrationSourceWs) Do(state *state.State, migrateOp *operations.Operati
 	if s.instance.Type() == instancetype.VM {
 		blockSize, err := storagePools.InstanceDiskBlockSize(pool, s.instance, migrateOp)
 		if err != nil {
-			return fmt.Errorf("Failed getting source disk size: %w", err)
+			return abort(fmt.Errorf("Failed getting source disk size: %w", err))
 		}
 
-		logger.Debugf("Set migration offer volume size for %q: %d", s.instance.Name(), blockSize)
+		l.Debug("Set migration offer volume size for %q: %d", logger.Ctx{"blockSize": blockSize})
```
Comment (monstermunchkin) [inline lxd/migrate_instance.go:467]: `%q` and `%d` can be removed as they aren't used

File: lxd/migrate_instance.go [hunk 2]
```diff
@@ -480,22 +495,10 @@ func (s *migrationSourceWs) Do(state *state.State, migrateOp *operations.Operati
...
-		s.sendControl(err)
-		return err
-	}
-
 	rsyncBwlimit = pool.Driver().Config()["rsync.bwlimit"]
 	migrationTypes, err = migration.MatchTypes(respHeader, migration.MigrationFSType_RSYNC, poolMigrationTypes)
 	if err != nil {
-		logger.Errorf("Failed to negotiate migration type: %v", err)
+		l.Error("Failed to negotiate migration type", logger.Ctx{"err": err})
 		return abort(err)
```
Comment (monstermunchkin) [inline lxd/migrate_instance.go:500]: The `abort` function already creates a log entry. We should do
```go
return abort(fmt.Errorf("Failed to negotiate migration type: %w", err))
```

File: lxd/migrate_instance.go [hunk 3]
```diff
@@ -729,20 +732,19 @@ func (s *migrationSourceWs) Do(state *state.State, migrateOp *operations.Operati
 	msg := migration.MigrationControl{}
 	err = s.recv(&msg)
 	if err != nil {
-		s.disconnect()
-		return err
+		return abort(err)
```
Comment (monstermunchkin) [inline lxd/migrate_instance.go:743]: Just for my understanding: the `abort` function doesn't call `s.disconnect()`. Will that cause issues?

File: lxd/storage/backend_lxd.go [hunk 4]
```diff
@@ -1776,10 +1808,29 @@ func (b *lxdBackend) CreateInstanceFromMigration(inst instance.Instance, conn io
...
+				if srcSnap.Name == snapName {
+					snapConfig = srcSnap.Config
+					snapDescription = srcSnap.Description
+
+					if srcSnap.ExpiresAt != nil {
+						snapExpiryDate = *srcSnap.ExpiresAt
+					}
+
+					break
+				}
```
Comment (monstermunchkin) [inline lxd/storage/backend_lxd.go:1828]: We could reduce the indentation here:
```go
for _, srcSnap := range srcInfo.Config.VolumeSnapshots {
	if srcSnap.Name != snapName {
		continue
	}

	snapConfig = srcSnap.Config
	snapDescription = srcSnap.Description

	if srcSnap.ExpiresAt != nil {
		snapExpiryDate = *srcSnap.ExpiresAt
	}

	break
}
```

**Review by monstermunchkin — CHANGES_REQUESTED** [review]

File: lxd/instance/drivers/driver_lxc.go [hunk 5]
```diff
@@ -210,11 +210,11 @@ func lxcCreate(s *state.State, args db.InstanceArgs) (instance.Instance, revert.
 
 	_, rootDiskDevice, err := d.getRootDiskDevice()
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, fmt.Errorf("Failed getting root disk: %w", err)
 	}
 
 	if rootDiskDevice["pool"] == "" {
-		return nil, nil, fmt.Errorf("The container's root device is missing the pool property")
+		return nil, nil, fmt.Errorf("The instances's root device is missing the pool property")
```
[the remark this answers is not in the document]
Comment (tomponline) [reply lxd/instance/drivers/driver_lxc.go:217]: It should be `instance's`, but it is wrong so will fix.

File: lxd/migrate_instance.go (hunk 1 above)
Comment (tomponline) [reply lxd/migrate_instance.go:467]: Good spot thanks!

File: lxd/storage/backend_lxd.go (hunk 4 above)
Comment (tomponline) [reply lxd/storage/backend_lxd.go:1828]: Will do.

**Comment by tomponline** [comment]
@monstermunchkin all points addressed.

**Review by monstermunchkin — APPROVED** [review]

**Comment by tomponline** [comment]
@stgraber ready for review.

**Pull request merged**